### DISCUSSION TITLE
SFRA currentCustomer.raw is null

### DISCUSSION THREAD

**Discussion by zoran-surlan** [SalesforceCommerceCloud/sfcc-ci#408] [Team Posts]
Hi all,
we are experiencing an issue with currentCustomer.raw profile is returned as null if we use Server.Replace with prep end and append functionalities. 
Anybody else experiencing something similar?

**Comment by zlatinz** [comment]
Hi, when you use server.replace you are practically replacing everything from the corresponding controller route with the new route you are providing. Anything that is prepared in the SFRA base cartridge controller route will not be accessible.

What exactly are you trying to customise, is it a controller or a model? Can you put the full path (relative from cartridge) to the file you are overriding?

**Comment by zoran-surlan** [comment]
We are trying to replace Account-SubmitRegistration and we have prepend methods from plugin_wishlist.
Yes we aware that Server.replace replaces entire request chain so we have added wishlist prepend again in our controller, so that request chain is fully restored, but for some reason req.currentCustomer.raw is null in the prepend method copied to our Account controller.
I have tried to debug through request constructor, but this part for defining customer is not accessible to my debugger because Object.defineProperty is used.

**Comment by zlatinz** [comment]
I am not sure what you mean by "we have prepend methods from plugin_wishlist", can you elaborate on this, or even better share a code snippet of what you are doing?

**Comment by zoran-surlan** [comment]
Here is the code snippet for the controller function that we are replacing, (all methods are part of the single controller):
```
server.replace('SubmitRegistration', server.middleware.https, csrfProtection.validateAjaxRequest, function (req, res, next) {
    const CustomerMgr = require('dw/customer/CustomerMgr');
    const Resource = require('dw/web/Resource');
...........

server.prepend('SubmitRegistration', function (req, res, next) {
    const viewData = res.getViewData();
    const list = productListHelper.getList(req.currentCustomer.raw, { type: 10 }); // here we get NPE
    viewData.list = list;
    res.setViewData(viewData);
    next();
});

```

**Comment by zlatinz** [comment]
Precisely.
So with replace you are getting the following chain:
[ server.middleware.https, csrfProtection.validateAjaxRequest, {your-anonymous-inline-function} ]
after prepend you are getting
[ {your-anonymous-inline-function-with-NPE} , server.middleware.https, csrfProtection.validateAjaxRequest, {your-anonymous-inline-function} ]
There is nothing called before your function with NPE to populate this property in req.
As apparently you are trying to get the Script API object that is for current customer, why not use instead of `req.currentCustomer.raw` just `customer` (which is static member of TopLevel.global class, see https://documentation.b2c.commercecloud.salesforce.com/DOC1/topic/com.demandware.dochelp/DWAPI/scriptapi/html/api/class_TopLevel_global.html)

**Comment by zoran-surlan** [comment]
I tried your suggestion Zlatin but without luck.
I also tried session.customer and again it is null.
This shouldn't happen in any case.

**Comment by challmannsf** [comment]
@zoran-surlan 
you are saying you are replacing `SubmitRegistration` , what does it look like in you implementation exactly?
The point of interest could be in ` this.on('route:BeforeComplete'` where the customer is authenticated.

**Comment by zoran-surlan** [comment]
@challmannsf 
yes we are replacing SubmitRegistration request chain.
BeforeComplete point isn't reached because our prepend method breaks due to NPE.
NPE is caused by req.currentCustomer.raw being null, I have tried to replace this customer with SFCC global objects for customer and customer from global session but with same result. 
As you know SFCC Session class for storefront session should never return a null for the customer.

**Comment by zlatinz** [comment]
In what context are you running this code?
Can you check the value of `session`, `request` and `response` objects?
Can you check the same code on different instance to see if it is instance specific issue?

**Comment by zoran-surlan** [comment]
Context is Customer Create Account page on storefront.
I already did check session, request objects and response view data as well as you can see in my last response to your questions.
Since this morning this same code is working fine on one sandbox but on others, I have raised support case to investigate deeper.
